### ISSUE TITLE
Corner case (?): tab.paired only diseased (or nondiseased)

### ISSUE THREAD

**Issue by anddis** [chstock/DTComPair#8]
Hello Christian,

I have been using DTComPair a lot to analyse real data from two paired studies. It's been extremely helpful!

I have stumbled upon what could be considered a corner-case: if all subjects are diseased (`d=1`) or non-diseased (`d=0`), `tab.paired` stops with an error:
```
Error in eval(substitute(d), data, parent.frame()) : 
  numeric 'envir' arg not of length one
```

What do you think about changing the check
```
if (identical(sort(unique(d)), as.integer(c(0, 1))) == FALSE)
```
with something like 
```
if (all(unique(d) %in% c(0, 1)) == FALSE)
```
?

(In a similar fashion, one could consider modifying the checks for `y1` and `y2`).

I am a bit hesitant because this would cause errors downstream (eg, in `sesp.rel`, but not in `tpffpf.rel`), and making the necessary changes to all functions would be a lot of work. At the same time, for example in the example below, calculating the rTPF would be ok...

```
> df <- data.frame(
+   d = rep(1, 10),
+   y1 = rbinom(10, 1, 0.3),
+   y2 = rbinom(10, 1, 0.7)
+ )
> paired.data <- tab.paired2(d, y1, y2, df)
> paired.data
Two binary diagnostic tests (paired design)

Test1: 'y1'
Test2: 'y2'

Diseased:
           Test1 pos. Test1 neg. Total
Test2 pos.          2          6     8
Test2 neg.          0          2     2
Total               2          8    10

Non-diseased:
           Test1 pos. Test1 neg. Total
Test2 pos.          0          0     0
Test2 neg.          0          0     0
Total               0          0     0

> tpffpf.rel(paired.data)
$tpf
       rel.tpf se.log.rel.tpf    lcl.rel.tpf    ucl.rel.tpf   pval.rel.tpf 
    4.00000000     0.61237244     1.20450229    13.28349491     0.02358585 

$fpf
       rel.fpf se.log.rel.fpf    lcl.rel.fpf    ucl.rel.fpf   pval.rel.fpf 
           NaN            NaN            NaN            NaN            NaN 

$alpha
[1] 0.05
```

(`tab.paired2` is a modified version of `tab.paired` where I made the suggested change above to the `if` condition).

What do you think? 
Thank you for taking the time to consider this.

Andrea

**Comment by chstock** [comment]
Hi @anddis,

thanks for pointing this out!

The error message above clearly isn't user friendly. I seem to get a slightly different message, which however isn't ideal too.

```r
df <- data.frame(
  d = rep(1, 10),
  y1 = rbinom(10, 1, 0.3),
  y2 = rbinom(10, 1, 0.7)
  )
paired.data <- tab.paired(d, y1, y2, df)
# Error in tab.paired(d, y1, y2, df) : 
#   Disease status (d) must be coded as 0 or 1.
```

To me it looks like it it may be good to add your proposal
```r
if (all(unique(d) %in% c(0, 1)) == FALSE)
```
after

```r
if (identical(sort(unique(d)), as.integer(c(0, 1))) == FALSE)
```
plus a sensible error message. 
I would keep the current check to capture other coding errors.

I am generally in favor of using [assertions](https://www.[r-project.org/nosvn/pandoc/assertthat.html](https://www.r-project.org/nosvn/pandoc/assertthat.html)), but this wasn't around when the code was developed and it would be cumbersome to change all functions. Nevertheless, feel free to use [asserthat ](https://cran.r-project.org/web/packages/assertthat/index.html) as an imported package.

I don't think it is necessary to change the checks for `y1` and `y2`.

What you like to make a proposal? (either a code snippet here or a PR)

Best, 
Christian

**Comment by anddis** [comment]
Thanks for your swift reply, Christian.

If we want `tab.paired` to deal with (and allow for) data where all subjects are either diseased (`d=1`) or nondiseased (`d=0`), my suggestion would be to **replace**

```r
if (identical(sort(unique(d)), as.integer(c(0, 1))) == FALSE)
```
with
```r
if (all(unique(d) %in% as.integer(c(0, 1))) == FALSE)
```

The existing `if` condition is **stricter** than the one I've proposed (both 0's and 1's must be present in the data _versus_ either only 0's or only 1's or both can appear in the data). In the example above (data `df` with only diseased subjects), the existing condition will stop the execution of the code, the alternative condition won't. Adding my proposal after the existing condition will stop the execution of the code as well.

Apologies if my initial message was unclear.

Two notes:
- Again, imho, it makes sense for `tab.paired` to allow for data where only diseased or nondiseased are present. The consequences are, however, even more uninformative error messages from some of the package's functions downstream. That's why I asked your opinion about the ultimate usefulness of this change. For example `acc.paired`:
```r
> df <- data.frame(
+   d = rep(1, 10),
+   y1 = rbinom(10, 1, 0.3),
+   y2 = rbinom(10, 1, 0.7)
+ )
> paired.data <- tab.paired2(d, y1, y2, df) # modified version of tab.paired without checks on d
> acc.paired(paired.data)
Error in if (spec.lcl < 0) spec.lcl <- 0 : 
  missing value where TRUE/FALSE needed
```
- You are right, I copy/pasted the wrong error message in my initial post. Apologies. The error message in your reply is what I am also getting:
```r
Error in tab.paired(d, y1, y2, df) : 
  Disease status (d) must be coded as 0 or 1.
```

Thanks again for your patience.

**Comment by chstock** [comment]
...right, sorry, I now got why it is stricter and I believe I better understand your point.

I agree that `tab.paired` should allow for data where everyone is diseased or everyone is non-diseased. It primarily serves as a way to aggregate and to display a dataset, which should be doable also for this special case. I'd tend to see the downstream issues as the "problems" of the downstream functions. So it would be fine from my perspective to modify `tab.paired` according to your proposal. 

Regarding modifications to the downstream functions, I don't have a strong opinion at the moment. Certainly, addressing the issues arising from this corner case and more user-friendliness in error messages is desirable, but I don't think it's absolute necessary to implement this right now. Please feel free though to tackle this issue too.

Hope this makes sense.

**Comment by anddis** [comment]
It definitely make sense and I 100% agree with everything you wrote.

For the same reason you pointed out (_"It primarily serves as a way to aggregate and to display a dataset..."_), you may want to consider replacing the existing check for `y1` and `y2` too (lines 192–195 https://github.com/chstock/DTComPair/blob/main/R/DTComPair.R) with something analogous to what we discussed, ie:

```r
if (all(unique(y1) %in% as.integer(c(0, 1))) == FALSE)
```
and
```r
if (all(unique(y2) %in% as.integer(c(0, 1))) == FALSE)
```

Lastly, `tab.1test`'s checks for `d` and `y` (lines 23–26 https://github.com/chstock/DTComPair/blob/main/R/DTComPair.R) could also be amended, for the same reason above.

**Comment by chstock** [comment]
The changes have now been implemented in dev version 1.2.3.

Now:

```r 
> df <- data.frame(
+      d = rep(1, 10),
+      y1 = rbinom(10, 1, 0.3),
+      y2 = rbinom(10, 1, 0.7)
+    )
> tab.paired(d, y1, y2, df)
Two binary diagnostic tests (paired design)

Test1: 'y1'
Test2: 'y2'

Diseased:
           Test1 pos. Test1 neg. Total
Test2 pos.          4          4     8
Test2 neg.          1          1     2
Total               5          5    10

Non-diseased:
           Test1 pos. Test1 neg. Total
Test2 pos.          0          0     0
Test2 neg.          0          0     0
Total               0          0     0

> tab.1test(d = df$d, y = df$y1) 
Binary diagnostic test 'df$y1'

          Diseased Non-diseased Total
Test pos.        5            0     5
Test neg.        5            0     5
Total           10            0    10
```

**Comment by chstock** [comment]
The issue should be resolved.
